python/mxnet/contrib/autograd.py

```diff
@@ -10,15 +10,70 @@
 from ..ndarray import NDArray, zeros_like
 from ..symbol import _GRAD_REQ_MAP
 
-def set_recording(recording):
-    """Turn on or turn off operator recording.
+
+def set_is_training(is_train):
+    """Set status to training/not training. When training, graph will be constructed
+    for gradient computation. Operators will also run with ctx.is_train=True. For example,
+    Dropout will drop inputs randomly when is_train=True while simply passing through
+    if is_train=False.
 
     Parameters
     ----------
-    recording: bool
+    is_train: bool
+
+    Returns
+    -------
+    previous state before this set.
+    """
+    prev = ctypes.c_int()
+    check_call(_LIB.MXAutogradSetIsTraining(
+        ctypes.c_int(is_train), ctypes.byref(prev)))
+    return bool(prev.value)
+
+
+class TrainingStateScope(object):
+    """Scope for managing training state.
+
+    Example::
+        with TrainingStateScope(True):
+            y = model(x)
+            compute_gradient([y])
     """
-    check_call(_LIB.MXAutogradSetRecording(
-        ctypes.c_int(recording)))
+    def __init__(self, enter_state):
+        self._enter_state = enter_state
+        self._prev = None
+
+    def __enter__(self):
+        self._prev = set_is_training(self._enter_state)
+
+    def __exit__(self, ptype, value, trace):
+        if self._prev != self._enter_state:
+            set_is_training(self._prev)
+
+
+def train():
```

Comment:
`train` and `test` sound too generalized, maybe consider `train_scope` and `test_scope`